# Simulado Completo — Prova Final de Programação (Fundamentos)

**Gerado em 09.06 para a final de 10.06.** Espelha o formato das provas do Hitoshi (FinalFunProg 2025.1 / 2025.2 / simulado) + a camada de análise de dados da atividade em grupo (AGP).

## ATENÇÃO (regras da prova real, valem aqui)
* Os códigos desenvolvidos devem servir para **quaisquer cenários**, não apenas para os exemplos informados.
* Não remover `#` das áreas de código; completar somente após o `:`.
* Na Turma utilizar somente **AE1, AE2, AE3, AE4, AE5, AE6**.

## Como usar
1. Resolva de cima a baixo, **cheat sheet fechado**, cronometrando (tempo alvo em cada questão).
2. Cada questão de código tem linhas **✅ Confira** com a saída esperada (verificada nos dados reais): rode seu código e compare.
3. Gabarito completo em `SimuladoFinalCompleto_GABARITO.ipynb`. Só abra depois de terminar a questão (ou se travar 10+ min).
4. Parte A = formato exato da final individual. Parte B = análise estilo AGP (o professor sinalizou prova mista).

---

**Questão 0 (identificação):**<br>
Preencha seus dados na célula abaixo. Veja exemplo:
```python
# Nome: Getúlio Vargas
# Turma: AE1
# Código: C00000000
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 0 - NÃO DELETAR ESTA LINHA !!

# Nome: Arthur Gomes Malucelli
# Turma: AE6
# Código: C3023369


---
# PARTE A — Formato da prova final individual

**Questão 1A (1,0 pt)** — *conceito, tempo alvo: 5 min*

```python
1 preco = input("Qual o preço do produto? ")
2 desconto = 10
3 preco_final = preco - desconto
4 print("Preço final:", preco_final)
```

Este programa gera um erro. \
Em qual linha o erro é gerado, por que, e como fazer para consertá-lo?

Responda justificando sua resposta.

CRITÉRIO DE CORREÇÃO (1,0): \
Não há pontuação parcial nessa questão

✅ Confira: sua resposta precisa apontar UMA linha específica, citar o TIPO de dado que `input()` retorna e dar o conserto.

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 1A - NÃO DELETAR ESTA LINHA !!

Erro na linha 1, input é sempre str, nao tem como tirar um numero de um str, pra concertar deve se adicionar int(input"Qual o preço do produto?"))


**Questão 1B (1,0 pt)** — *conceito, tempo alvo: 5 min*

Um caixa eletrônico tem duas rotinas:

1. Imprimir o extrato com os **últimos 10 lançamentos** da conta.
2. Liberar notas de R\$50, uma por vez, **enquanto o saldo restante permitir** mais um saque de R\$50.

Ambos os laços podem ser realizados com *for* ou *while*. Contudo, existe uma opção mais apropriada para cada caso.

Entre *for* e *while*, qual comando você usaria em cada rotina? Justifique.

CRITÉRIO DE CORREÇÃO (1,0): \
Não há pontuação parcial nessa questão

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 1B - NÃO DELETAR ESTA LINHA !!

**Questão 1C (1,0 pt)** — *conceito, tempo alvo: 5 min*

Considere que `clientes` é um dataframe com várias colunas, entre elas **"Cidade"** (texto) e **"Limite"** (número).

Precisamos aumentar em 15% o limite de todos os clientes de "Campinas". \
**Exatamente duas** das opções abaixo estão corretas. Assinale todas as alternativas corretas \
(não tente rodar o código, porque `clientes` não está definido):

```python
a) clientes.loc[clientes["Cidade"]=="Campinas", "Limite"] *= 1.15
b) clientes[clientes["Cidade"]=="Campinas"]["Limite"] = clientes[clientes["Cidade"]=="Campinas"]["Limite"] * 1.15
c) clientes.loc["Cidade"=="Campinas", "Limite"] = clientes["Limite"] * 1.15
d) clientes.loc[clientes["Cidade"]=="Campinas", "Limite"] = clientes.loc[clientes["Cidade"]=="Campinas", "Limite"] * 1.15
e) clientes[clientes[Cidade]=="Campinas", "Limite"] *= 1.15
```

CRITÉRIO DE CORREÇÃO (1,0): \
Não há pontuação parcial nessa questão

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 1C - NÃO DELETAR ESTA LINHA !!
A - Usa a funcao loc para localizar no dataframe clientes, onde a coluna Cidade tem valor == Campinas, puxa o limite dessa coluna, e o redefine usando =* 1.15
D - Mesma forma da A, usa a função LOC, para localizar as colunas, mas faz de forma separada, primeiro, ela acha nos Clientes aonde voce tem a coluna Cidade com valor Campinas, e puxa os Limites dali, dps redefine os limites puxados como 15% a mais
# alternativas corretas: ... porque ...

**Questão 2A (1,0 pt)** — *ler pandas, tempo alvo: 5 min*

Analise o programa abaixo, e considere que foi executado sem erros.
O dataframe `vendas` contém uma linha para cada pedido.
Colunas `Cliente` e `Total` contêm o nome do cliente e o valor do pedido em reais.

```python
import pandas as pd
URL = "https://exemplo.com/vendas.xlsx"
vendas = pd.read_excel(URL)

grandes = vendas.loc[vendas["Total"] >= 5000, "Cliente"]

print(len(grandes))
```
O que este programa faz?

Responda e justifique sua resposta.

CRITÉRIO DE CORREÇÃO (1,0): \
Não há pontuação parcial nessa questão

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 2A - NÃO DELETAR ESTA LINHA !!
 Esse programa importa a planilha vendas.xlsx, e define grandes, como lugares onde o total de vendas p este cliente é maior ou igual a 5k, apos isso, imprimi quantos clientes grandes voce tem.

**Questão 2B (1,0 pt)** — *ler pandas, tempo alvo: 5 min*

```python
import pandas as pd
estoque = pd.read_excel("estoque.xlsx")

estoque.loc[estoque["Categoria"] == "Bebidas", "Preço"] = estoque["Preço"] * 0.85

print(estoque.head())
```
O que este programa faz?

OBS: não tente rodar esse programa.

CRITÉRIO DE CORREÇÃO (1,0): \
São três blocos de instruções. Cada bloco vale 0,33 pt

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 2B - NÃO DELETAR ESTA LINHA !!

 Este programa define o df estoque como a planilha do excel estoque.xlxs, depois, usa .loc para localizar neste df aonde na coluna Categoria, voce tem o valor Bebidas, apos isso, redefine o Preco dessas bebidas com um desconto de 15%, e imprimi os primeiros 5 linhas da coluna.

**Questão 3 (2,5 pts)** — *input + validação + condicional, tempo alvo: 20 min*

A taxa de entrega do restaurante Dr. Paçoca Delivery depende da **distância** até o cliente e do **período** do dia.

* Se a distância for até 2.000 m: R\$5,00 fixo
* Se a distância estiver entre 2.001 m e 8.000 m: R\$1,20 por quilômetro (`distância em metros / 1000` × tarifa)
* Se a distância estiver entre 8.001 m e 15.000 m: R\$1,00 por quilômetro
* Se a distância for acima de 15.000 m: R\$25,00 fixo, independente da distância

Se o período for "D" (dia): sem adicional \
Se o período for "N" (noite): **dobra** o valor da taxa (inclusive as taxas fixas)

Você deve criar um programa que: \
a) receba do usuário a distância em metros (valor inteiro) e o período ("D" ou "N") \
b) valide as informações:
   - a distância deve ser um número inteiro (não pode receber caracteres não numéricos, inclusive ponto decimal)
   - o período deve ser apenas "D" ou "N". Se o usuário digitar "d" ou "n", o programa deve reconhecer (maiúsculas)
   - Se houver erro, mostrar mensagem `ERRO DE ENTRADA` e não calcular a taxa

c) caso não tenha erros, apresente o valor da taxa.

DICAS:
* `"123".isdigit()` retorna `True`; `"12.3".isdigit()` e `"abc".isdigit()` retornam `False`
* o método `.upper()` transforma `"n"` em `"N"`

CRITÉRIO DE CORREÇÃO (2,5): \
1,0: o programa coletou os dados, validou as entradas e realizou as transformações necessárias \
1,5: o programa calcula corretamente a taxa em todos os cenários e a imprime

OBS: Não há pontuação parcial dentro de cada rubrica.

✅ Confira (saídas verificadas):
* entrada `1500` + `d` → `Taxa de entrega: R$ 5.00`
* entrada `9000` + `N` → `Taxa de entrega: R$ 18.00`
* entrada `25000` + `D` → `Taxa de entrega: R$ 25.00`
* entrada `12.5` + `D` → `ERRO DE ENTRADA`
* entrada `3000` + `X` → `ERRO DE ENTRADA`

# ESPAÇO PARA RESPOSTA DA QUESTAO 3 - NÃO DELETAR ESTA LINHA !!

distancia = input("Digite a distância em metros do seu pedido")
periodo = input("Digite o periodo do seu pedido")
periodo = periodo.upper()
if periodo not in ["D", "N"]:
    print("ERRO DE ENTRADA")
else:
    if distancia.isdigit():
        distancia = int(distancia)
        if periodo == "D":
            taxa = 1
        else:
            taxa = 2
        if distancia < 2000:
            delivery = 5 * taxa
        elif distancia < 8000:
            delivery = distancia / 1000 * 1,2 * taxa
        elif distancia < 15000:
            delivery = distancia / 1000 * 1 * taxa
        else:
            delivery = 25 * taxa
        print(delivery)
    else:
        print("ERRO DE ENTRADA")


In [81]:
distancia = input("Digite a distância em metros do seu pedido")
periodo = input("Digite o periodo do seu pedido")
periodo = periodo.upper()
if periodo not in ["D", "N"]:
    print("ERRO DE ENTRADA")
else:
    if distancia.isdigit():
        distancia = int(distancia)
        if periodo == "D":
            taxa = 1
        else:
            taxa = 2
        if distancia <= 2000:
            delivery = 5 * taxa
        elif distancia <= 8000:
            delivery = distancia / 1000 * 1.2 * taxa
        elif distancia <= 15000:
            delivery = distancia / 1000 * 1 * taxa
        else:
            delivery = 25 * taxa
        print(delivery)
    else:
        print("ERRO DE ENTRADA")

Digite a distância em metros do seu pedido 8000
Digite o periodo do seu pedido d


9.6


**Questão 4A (2,5 pts)** — *DataFrame: trocar valor + faturar, tempo alvo: 12 min*

O arquivo carregado na célula abaixo possui a lista de pedidos do Mr. Paçoca. A empresa fará uma promoção: todos os pedidos de um determinado **Estado** terão o preço unitário multiplicado por um **fator** de desconto (ex: fator 0.9 = 10% de desconto).

Crie um programa em que, após o carregamento do dataframe: \
a) solicite ao usuário a sigla do Estado e o fator (suponha que ele vá inserir os dados corretamente; aceite a sigla em minúsculas) \
b) aplique o fator ao preço unitário **apenas** dos pedidos daquele Estado, utilizando obrigatoriamente as variáveis definidas em a) \
c) crie uma nova coluna com o total de cada pedido (preço unitário × quantidade) \
d) informe o faturamento global de todos os clientes

**ATENÇÃO:** se você receber um alerta de `FutureWarning / ChainedAssignmentError`, ignore-o.

CRITÉRIO DE CORREÇÃO (2,5): \
0,5: o programa coleta os inputs (item a) \
1,0: com os inputs, troca corretamente o preço unitário dos pedidos selecionados (item b) \
1,0: o programa informa o faturamento global (itens c e d)

✅ Confira: com Estado `sp` e fator `0.9`, a saída é `Faturamento global: R$ 2158282.17`

In [94]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 4A - NÃO DELETAR ESTA LINHA !!

import pandas as pd
pacoca = pd.read_excel("MrPacoca.xlsx")

In [95]:
pacoca.head()

,Data,Dia,Mês,Ano,Tipo,Cliente,Código Cliente,Quantidade,Preço Unitário,Estado,Nome da embalagem,Embalagem,Peso por embalagem,Código do Produto,Peso unitário
0,2022-11-25,25,11,2022,C,mostardinha,9813423,9740,0.23,MG,Pote,163,150,163C0150,15
1,2022-07-27,27,7,2022,C,azeitoninha,7104250,12640,0.23,TO,Pote,163,150,163C0150,15
2,2022-08-18,18,8,2022,C,aipinho,1503649,7010,0.23,BA,Pote,163,150,163C0150,15
3,2022-04-23,23,4,2022,C,limãozinho,2613386,3810,0.23,SP,Pote,163,150,163C0150,15
4,2022-08-14,14,8,2022,C,batatinha,6498768,7910,0.23,SP,Pote,163,150,163C0150,15


In [96]:
estado = input("Estado para dar Desconto:")
estado = estado.upper()
fator = float(input("Fator do desconto:"))
pacoca.loc[pacoca["Estado"]== estado, "Preço Unitário"] *= fator
pacoca["Total Pedido"] = pacoca["Preço Unitário"] * pacoca["Quantidade"]
faturamento_global = pacoca["Total Pedido"].sum()
print(faturamento_global)

Estado para dar Desconto: sp
Fator do desconto: 0.9


2158282.1680000005


In [90]:
pacoca.head()

,Data,Dia,Mês,Ano,Tipo,Cliente,Código Cliente,Quantidade,Preço Unitário,Estado,Nome da embalagem,Embalagem,Peso por embalagem,Código do Produto,Peso unitário,Total Pedido
0,2022-11-25,25,11,2022,C,mostardinha,9813423,9740,0.230,MG,Pote,163,150,163C0150,15,2240.20
1,2022-07-27,27,7,2022,C,azeitoninha,7104250,12640,0.230,TO,Pote,163,150,163C0150,15,2907.20
2,2022-08-18,18,8,2022,C,aipinho,1503649,7010,0.230,BA,Pote,163,150,163C0150,15,1612.30
3,2022-04-23,23,4,2022,C,limãozinho,2613386,3810,0.207,SP,Pote,163,150,163C0150,15,788.67
4,2022-08-14,14,8,2022,C,batatinha,6498768,7910,0.207,SP,Pote,163,150,163C0150,15,1637.37


**Questão 4B (2,5 pts)** — *DataFrame: derivar coluna + reajuste + diferença, tempo alvo: 15 min*

O arquivo carregado na célula abaixo possui a lista de pedidos de 2025. Os pedidos de um determinado **estado** sofrerão um reajuste de preço, e a empresa quer saber a diferença de faturamento.

Crie um programa em que: \
a) o pandas calcule o preço unitário de cada pedido (coluna Total / Quantidade) e o inclua em uma nova coluna chamada `Preço Unitário` \
b) o pandas some o valor de `reajuste` ao preço unitário **apenas** dos pedidos do estado informado (coluna Estado = estado) \
c) o pandas calcule o novo total de cada pedido (preço unitário reajustado × quantidade) \
d) seja calculada e apresentada a diferença entre o faturamento original e o faturamento após o reajuste

Utilize as variáveis `reajuste` e `estado` conforme código abaixo. \
O programa deve funcionar para **quaisquer** valores informados nestas variáveis.

**ATENÇÃO:** se você receber um alerta de `FutureWarning / ChainedAssignmentError`, ignore-o.

CRITÉRIO DE CORREÇÃO (2,5): \
0,5: o programa calculou corretamente o preço unitário \
1,0: o programa filtra os pedidos e troca o valor unitário dos pedidos selecionados \
1,0: o programa calcula o novo total e informa a diferença

✅ Confira: com `reajuste = 0.02` e `estado = "MG"`, a saída é `Diferença de faturamento: R$ 18885.84`

In [124]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 4B - NÃO DELETAR ESTA LINHA !!

import pandas as pd
pedidos = pd.read_excel("pedidos.xlsx", index_col=0)

reajuste = 0.02
estado = "MG"
pedidos["Preço Unitário"] = pedidos["Total"] / pedidos["Quantidade"]
fat_antes = pedidos["Total"].sum()
pedidos.loc[pedidos["Estado"] == estado, "Preço Unitário"] += reajuste
pedidos["Total"] = pedidos["Preço Unitário"] * pedidos["Quantidade"]
fat_dps = pedidos['Total'].sum()
dif_fat = fat_dps - fat_antes
print(dif_fat)

18885.83999999985


**Questão 4C (2,0 pts)** — *DataFrame: busca por id + árvore de decisão, tempo alvo: 15 min*

Considere o arquivo `dataset_checkup_1000linhas.csv` (colunas: `id, idade, peso, altura, fumante, historico_familiar, diabetes, pressao_sistolica, pressao_diastolica`).

A classificação de **risco** de cada paciente segue as regras abaixo (nesta ordem de prioridade), onde IMC = peso / altura²:

1. Se `pressao_sistolica >= 140` **OU** `pressao_diastolica >= 90` → **consulta imediata**
2. Caso contrário, se `IMC >= 30` **E** `fumante == "sim"` → **risco alto**
3. Caso contrário, se `IMC >= 30` **OU** `idade >= 60` → **risco moderado**
4. Caso nenhuma anterior seja verdadeira → **risco baixo**

Monte um programa que: \
1. Carregue o dataset em um dataframe chamado `pacientes` (já feito na célula abaixo) \
2. Peça ao usuário um `id` (pode assumir que será um inteiro válido) \
3. Se o `id` **não existir**: apresente `Paciente não encontrado` \
4. Se existir: selecione a linha, calcule o IMC e apresente no formato **exato** (IMC com 1 casa decimal, inclusive o travessão `–`):

`Paciente 123 (IMC: 31.2) – Risco: risco alto`

CRITÉRIO DE CORREÇÃO (2,0): \
1,0: leu o dataset, verificou a existência do id e tratou o caso não encontrado \
1,0: aplicou corretamente TODAS as regras na ordem e exibiu no formato pedido

✅ Confira (saídas verificadas):
* id `894` → `Paciente 894 (IMC: 26.3) – Risco: consulta imediata`
* id `741` → `Paciente 741 (IMC: 32.4) – Risco: risco alto`
* id `1` → `Paciente 1 (IMC: 31.2) – Risco: risco moderado`
* id `223` → `Paciente 223 (IMC: 28.0) – Risco: risco baixo`
* id `1500` → `Paciente não encontrado`

In [138]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 4C - NÃO DELETAR ESTA LINHA !!

import pandas as pd
pacientes = pd.read_csv("dataset_checkup_1000linhas.csv")
pacientes.head()

,id,idade,peso,altura,fumante,historico_familiar,diabetes,pressao_sistolica,pressao_diastolica
0,1,23,96.5,1.76,não,não,sim,123,72
1,832,66,76.4,1.71,não,sim,não,132,84
2,223,59,83.7,1.73,não,não,não,132,84
3,813,45,74.1,1.78,sim,sim,não,114,80
4,795,45,72.4,1.62,não,não,não,117,75


In [149]:
input_id = int(input("ID:"))
if input_id in pacientes["id"].values:
    linha = pacientes[pacientes["id"] == input_id].iloc[0]
    ##atribui IMC ao paciente
    imc = linha["peso"] / linha["altura"] ** 2
    if linha["pressao_sistolica"] >= 140 or linha["pressao_diastolica"] >= 90:
        risco = "consulta imediata"
    elif imc >= 30 and linha["fumante"] == "sim":
        risco = "risco alto"
    elif imc >= 30 and linha["idade"] >= 60:
        risco = "risco moderado"
    else:
        risco = "risco baixo"
    print(f"Paciente {input_id} (IMC: {imc:.1f}) – Risco: {risco}")
else:
    print("Paciente não encontrado")

ID: 894


Paciente 894 (IMC: 26.3) – Risco: consulta imediata


**Questão 5A (3,0 pts)** — *parse de strings em laço, tempo alvo: 15 min*

A Dr. Paçoca recebe as notas fiscais das lojas em formato de string, com valores separados por uma barra vertical (`|`):

```python
"LOJA03|98214|159.9"
```

O primeiro item é o código da loja, o segundo é o número da nota e o terceiro é o valor em reais. As notas chegam várias de uma vez, em uma lista:

```python
registros = ["LOJA03|98214|159.9", "LOJA01|98215|89.5",
             "LOJA12|98216|1230.0", "LOJA03|98217|45.25"]
```

Escreva um programa que mostre cada nota no formato:

**Nota {número} ({loja}) - R\$ {valor}**

O programa também deverá imprimir, ao final, a quantidade de notas e o total dos valores **com 2 casas decimais**, no formato `N notas - Total: R$ X.XX`.

CRITÉRIO DE CORREÇÃO (3,0): \
0,5: o programa faz uma varredura sobre os registros via um laço \
1,0: o programa separa as informações de cada registro \
0,5: o programa imprime as informações no modelo solicitado \
1,0: o programa imprime uma única linha totalizadora

ATENÇÃO: O programa deverá funcionar para qualquer lista, não apenas para o exemplo.

✅ Confira (saída verificada para a lista do exemplo):
```
Nota 98214 (LOJA03) - R$ 159.9
Nota 98215 (LOJA01) - R$ 89.5
Nota 98216 (LOJA12) - R$ 1230.0
Nota 98217 (LOJA03) - R$ 45.25

4 notas - Total: R$ 1524.65
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 5A - NÃO DELETAR ESTA LINHA !!

registros = ["LOJA03|98214|159.9", "LOJA01|98215|89.5",
             "LOJA12|98216|1230.0", "LOJA03|98217|45.25"]



**Questão 5B (3,0 pts)** — *parse de datas em laço, tempo alvo: 20 min*

Temos uma lista de registros de data/hora em formato string, sendo que alguns contêm somente a data. Os dois formatos possíveis:

* `DD/MM/YYYY HH:MM`, quando há data e hora
* `DD/MM/YYYY`, quando há somente a data

```python
datas = ["25/12/2025 01:30", "01/01/2026", "15/03/2026 14:00",
         "07/09/2025 08:01", "30/06/2026 23:45", "12/10/2025"]
```

Construa um programa que imprima uma linha para cada elemento: dia (número) + mês (**por extenso**) + ano + a parte da hora/minuto, se houver.

Regras:
- Se a hora for igual a 1, escrever "à 1 hora"
- Se a hora for diferente de 1, escrever "às X horas"
- Se os minutos forem 0, não escrever a parte dos minutos
- Se os minutos forem 1, escrever "1 minuto"
- Se os minutos forem maiores que 1, escrever "X minutos"
- A parte dos minutos, quando existir, liga-se à hora pela palavra "e"

DICAS:
* `meses_do_ano = ['janeiro', 'fevereiro', 'março', 'abril', 'maio', 'junho', 'julho', 'agosto', 'setembro', 'outubro', 'novembro', 'dezembro']`
* `int('02')` retorna `2`
* crie a lógica para um registro, depois envolva num laço

CRITÉRIO DE CORREÇÃO (3,0): \
0,5: implementou um laço \
0,5: separou a data da hora (se houver) \
1,0: lógica correta para distinguir registros com e sem hora \
1,0: rearranjou os componentes para a resposta final correta

ATENÇÃO: deve funcionar para qualquer lista no formato apresentado.

✅ Confira (saída verificada para a lista do exemplo):
```
25 de dezembro de 2025 à 1 hora e 30 minutos
1 de janeiro de 2026
15 de março de 2026 às 14 horas
7 de setembro de 2025 às 8 horas e 1 minuto
30 de junho de 2026 às 23 horas e 45 minutos
12 de outubro de 2025
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO 5B - NÃO DELETAR ESTA LINHA !!

datas = ["25/12/2025 01:30", "01/01/2026", "15/03/2026 14:00",
         "07/09/2025 08:01", "30/06/2026 23:45", "12/10/2025"]



---
# PARTE B — Análise de dados (formato AGP / atividade em grupo)

O professor sinalizou que a final pode misturar programação básica com análise. Esta parte espelha a AGP de combustíveis, usando `pedidos.xlsx` + `custos_2025.csv` (custo unitário de cada produto por mês, **CSV brasileiro**: colunas separadas por `;` e decimal com `,`).

**As células compartilham estado: rode em ordem.** Setup abaixo já carrega `pedidos`.

*Tempo alvo da Parte B inteira: 50 min.*

In [ ]:
# SETUP — rodar antes de tudo
import pandas as pd
import numpy as np
pedidos = pd.read_excel("pedidos.xlsx", index_col=0)


**B1 (1,0 pt)** — Carregue `custos_2025.csv` num dataframe `custos`. Atenção ao separador e ao decimal. Mostre:
(i) o shape, (ii) a lista de colunas, (iii) os dtypes (a coluna `Custo Unitário` DEVE ser numérica, não texto).

✅ Confira:
```
Shape: (48, 3)
Colunas: ['Código do Produto', 'Mês', 'Custo Unitário']
Código do Produto        str
Mês                    int64
Custo Unitário       float64
dtype: object
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO B1 - NÃO DELETAR ESTA LINHA !!



**B2 (1,5 pts)** — Em `pedidos`: \
(i) conte quantos pedidos existem de cada `Nome da embalagem`; \
(ii) obtenha mínimo, média e máximo da coluna `Total` em **uma única operação**, arredondados a 2 casas.

📌 Três chamadas separadas (`.min()`, `.mean()`, `.max()`) não atendem ao enunciado.

✅ Confira:
```
Nome da embalagem
Pote       785
Display    150
Caixa      148
Name: count, dtype: int64

min       32.20
mean    2052.04
max     5034.72
Name: Total, dtype: float64
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO B2 - NÃO DELETAR ESTA LINHA !!



**B3 (1,0 pt)** — Os 3 primeiros caracteres do `Código do Produto` (ex: `163C0150` → `163`) deveriam corresponder à coluna `Embalagem`. Crie uma coluna `Prefixo` com esses 3 caracteres (sem laço, vetorizado) e mostre **quantas linhas** têm `Prefixo` diferente de `Embalagem`.

DICA: `Embalagem` é número; compare como texto usando `.astype(str)`.

✅ Confira:
```
Linhas com prefixo diferente da embalagem: 0
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO B3 - NÃO DELETAR ESTA LINHA !!



**B4 (2,0 pts)** — Construa um dataframe `resumo` com o valor **médio** de `Total` por `Estado` (linhas), com cada `Nome da embalagem` virando uma **coluna**. `Estado` deve ser coluna comum (não índice). Mostre o shape e as 5 primeiras linhas.

📌 Erro comum: esquecer o `reset_index()` e entregar índice em vez de coluna.

✅ Confira (NaN onde o estado não tem pedidos daquela embalagem — há 1 células NaN, é esperado):
```
Shape: (14, 4)
  Estado    Caixa  Display     Pote
0     AC  2949.75  1828.13  1771.24
1     AL  3859.00  3663.84  1772.54
2     AP  2978.86  4094.40  1844.84
3     BA  3171.71  1743.14  1753.08
4     GO  2776.50  4153.44  1010.32
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO B4 - NÃO DELETAR ESTA LINHA !!



**B5 (2,0 pts)** — Cruze `pedidos` com `custos` num dataframe `dados`, usando as **duas** chaves em comum. Depois: \
(i) mostre o shape após o merge (nenhuma linha deve se perder); \
(ii) crie `Preço Unitário` (= Total / Quantidade); \
(iii) crie `Margem Unitária` (= Preço Unitário − Custo Unitário); \
(iv) mostre min, média e máx da margem (uma operação, 4 casas) e **quantos pedidos têm margem negativa**.

📌 Faltar uma chave no merge multiplica linhas (cruzaria custo de um mês com pedido de outro).

✅ Confira:
```
Shape após merge: (1083, 18)
min    -0.0195
mean    0.0499
max     0.1276
Name: Margem Unitária, dtype: float64
Pedidos com margem negativa: 102
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO B5 - NÃO DELETAR ESTA LINHA !!



**B6 (2,5 pts)** — Para fechar: \
(i) crie `Margem Total` (= Margem Unitária × Quantidade); \
(ii) mostre a margem unitária **média por Estado**, ordenada da maior para a menor; \
(iii) mostre quantos pedidos cada estado tem; \
(iv) crie uma coluna `Situação`: "prejuízo" se a margem unitária for negativa, "lucro" caso contrário (sem laço), e conte cada situação.

**(v) INTERPRETATIVA** (responda em markdown ou comentário): o estado líder do ranking é confiável como "melhor mercado"? Olhe o número de pedidos dele. Cite pelo menos 2 fatores que esses dados NÃO capturam antes de uma decisão de negócio.

✅ Confira (itens i–iv):
```
Estado
MS    0.0605
TO    0.0581
GO    0.0569
PR    0.0567
SP    0.0519
RJ    0.0495
AL    0.0484
AC    0.0483
MG    0.0481
BA    0.0481
MT    0.0466
SC    0.0465
RR    0.0459
AP    0.0421
Name: Margem Unitária, dtype: float64

Pedidos por estado:
Estado
SP    307
RJ    199
MG    129
SC     83
BA     50
TO     48
AP     48
MT     47
AC     44
RR     42
PR     37
MS     20
AL     18
GO     11
Name: count, dtype: int64

Situação
lucro       981
prejuízo    102
Name: count, dtype: int64
```

In [ ]:
# ESPAÇO PARA RESPOSTA DA QUESTAO B6 - NÃO DELETAR ESTA LINHA !!



---
## Checklist antes de "entregar" (e antes da prova real)

- [ ] Rodou o notebook inteiro de cima a baixo sem erro? (`Kernel → Restart & Run All`)
- [ ] Q3: testou TODAS as faixas e os dois períodos? E os dois casos de erro?
- [ ] Q4: usou as variáveis dadas (não os números do exemplo)?
- [ ] Q5: funciona se você trocar a lista por outra?
- [ ] Conferiu cada ✅ com sua saída?

**Auto-avaliação:** Parte A vale 10 (como a prova real), Parte B vale 10. Se você fechou ≥8 em cada sem abrir o gabarito, está pronto. Erros em questão SEM pontuação parcial (Q1, rubricas de Q3/Q4) pesam dobrado na prova: revise esses primeiro.